garage/experiment/local_tf_runner.py

```diff
@@ -163,41 +169,132 @@ def obtain_samples(self, itr, batch_size):
         """Obtain one batch of samples.
 
         Args:
-            itr: Index of iteration (epoch).
-            batch_size: Number of steps in batch.
+            itr(int): Index of iteration (epoch).
+            batch_size(int): Number of steps in batch.
                 This is a hint that the sampler may or may not respect.
 
         Returns:
             One batch of samples.
 
         """
-        if self.n_epoch_cycles == 1:
+        if self.train_args['n_epoch_cycles'] == 1:
             logger.log('Obtaining samples...')
         return self.sampler.obtain_samples(itr, batch_size)
 
-    def save_snapshot(self, itr, paths=None):
+    def save(self, epoch, paths=None):
         """Save snapshot of current batch.
 
         Args:
-            itr: Index of iteration (epoch).
-            paths: Batch of samples after preprocessed.
+            itr(int): Index of iteration (epoch).
+            paths(dict): Batch of samples after preprocessed.
 
         """
         assert self.has_setup
 
-        logger.log("Saving snapshot...")
-        params = self.algo.get_itr_snapshot(itr)
+        logger.log('Saving snapshot...')
+
+        params = dict()
+        # Saves arguments
+        params['setup_args'] = self.setup_args
+        params['train_args'] = self.train_args
+
+        # Save states
         params['env'] = self.env
+        params['algo'] = self.algo
         if paths:
             params['paths'] = paths
-        snapshotter.save_itr_params(itr, params)
+        params['last_epoch'] = epoch
+
+        snapshotter.save_itr_params(epoch, params)
+
         logger.log('Saved')
 
+    def restore(self, snapshot_dir, from_epoch='last', resume_now=True):
+        """Restore experiment from snapshot.
+
+        Args:
+            snapshot_dir(str): Directory of snapshot.
+            from_epoch(str or int): The epoch to restore from.
+                Can be 'first', 'last' or a number.
+                Not applicable when snapshot_mode='last'.
+            resume_now(bool): Resume training immediately.
```

Comment:
Hmm this API is kind of strange. `restore` should either always start training or never start training.

What if you made your runner stateful (i.e. it contains everything needed to run train)? Then `runner.restore()` always returns nothing, and you an add `runner.resume()` which uses the arguments acquired during `restore()`. `runner.resume()` can be based on `runner.train()`.

Reply:
Added now. PTAL.